vhost_user_backend/vhost_user_net: Add multiple worker threads

bjzhjing opened the pull request:
Add multiple thread support in vhost-user-backend and vhost-user-net backend, which could run multiple handler threads to handle multiple queue in parallel, and to improve performance finally.

To address: 
#571 
#628

bjzhjing commented:
Performance test for vhost-user-net backend by running the following commands:

```
./cloud-hypervisor/target/release/vhost_user_net \
        --net-backend "sock=/tmp/vunet.sock,num_queues=4"

```
```
./cloud-hypervisor/target/release/cloud-hypervisor \
        --cpus boot=2 \
        --memory size=512M,file=/dev/shm   \
        --kernel /home/cathy/tiger_lake/linux/vmlinux \
        --console off \
        --serial tty \
        --cmdline "console=ttyS0 reboot=k panic=1 nomodules i8042.noaux i8042.nomux i8042.nopnp i8042.dumbkbd root=/dev/vda3" \
        --disk path=/home/cathy/clear-31890-kvm.img   \
        --net "mac=52:54:00:02:d9:01,ip=192.168.100.1,vhost_user=true,socket=/tmp/vunet.sock,num_queues=4" \
        --rng
```
Performance test result for vhost-user-net backend. 

> 2 queue pairs, 2 cpus, running 2 iperf3 session

```
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-30.00  sec  71.7 GBytes  20.5 Gbits/sec    0             sender
[  5]   0.00-30.00  sec  71.7 GBytes  20.5 Gbits/sec                  receiver
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-30.00  sec  71.5 GBytes  20.5 Gbits/sec    0             sender
[  5]   0.00-30.00  sec  71.5 GBytes  20.5 Gbits/sec                  receiver
```

> 1 queue pair, 2 vcpus, running 2 iperf3 session
```
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-30.00  sec  58.4 GBytes  16.7 Gbits/sec   80             sender
[  5]   0.00-30.00  sec  58.4 GBytes  16.7 Gbits/sec                  receiver
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-30.00  sec  58.0 GBytes  16.6 Gbits/sec  167             sender
[  5]   0.00-30.00  sec  58.0 GBytes  16.6 Gbits/sec                  receiver
```
(20.5 - 16.6) / 16.6 = 0.234, throughput improved by ~23.4%.

rbradford commented on src/bin/vhost_user_fs.rs:
```diff
@@ -70,7 +71,7 @@ impl convert::From<Error> for io::Error {
 
 struct VhostUserFsBackend<F: FileSystem + Send + Sync + 'static> {
-    mem: Option<GuestMemoryMmap>,
+    mem: Option<Arc<ArcSwap<GuestMemoryMmap>>>,
```
I don't understand why this change is necessary?

rbradford replied:
The memory never gets modified (maybe we should do that, @sboeuf - if we have a hotplug do we need to tell the vhost-user backends about a different memory map?) so you don't need to protect it with an Arc<Mutex<T>> or a Arc<ArcSwap<T>> just an Arc<T> should be viable.

rbradford commented on vhost_user_backend/src/lib.rs:
```diff
@@ -462,7 +486,7 @@ struct VhostUserHandler<S: VhostUserBackend> {
     memory: Option<Memory>,
     vrings: Vec<Arc<RwLock<Vring>>>,
-    worker_thread: Option<thread::JoinHandle<VringWorkerResult<()>>>,
+    worker_thread: Option<Vec<thread::JoinHandle<VringWorkerResult<()>>>>,
```
Don't need to use an `Option<Vec<T>>` as you can just use an empty vector for the equivalent of the None case.

rbradford replied:
And it's plural now.

rbradford commented on vhost_user_backend/src/lib.rs:
```diff
@@ -330,7 +376,7 @@ enum VringWorkerError {
 
 pub struct VringWorker {
-    epoll_fd: RawFd,
+    pub epoll_fd: RawFd,
```
Layering violation?

rbradford commented on vhost_user_backend/src/lib.rs:
```diff
@@ -453,7 +499,7 @@ type VhostUserHandlerResult<T> = std::result::Result<T, VhostUserHandlerError>;
 struct VhostUserHandler<S: VhostUserBackend> {
     backend: Arc<RwLock<S>>,
-    worker: Arc<VringWorker>,
+    worker: Vec<Arc<VringWorker>>,
```
s/worker/workers/

bjzhjing commented:
@rbradford I search the CI log but not find any failed case, any suggestion on this?

rbradford commented:
Pretty clear in the output to me, search for "panic":

https://cloud-hypervisor-jenkins.westus.cloudapp.azure.com/job/cloud-hypervisor/job/PR-785/6/consoleFull

```
thread 'vring_worker' panicked at 'index out of bounds: the len is 1 but the index is 1', src/bin/vhost_user_fs.rs:173:33
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:77
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:61
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1028
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1412
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:65
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:50
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:188
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:205
  10: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:464
  11: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:373
  12: rust_begin_unwind
             at src/libstd/panicking.rs:302
  13: core::panicking::panic_fmt
             at src/libcore/panicking.rs:139
  14: core::panicking::panic_bounds_check
             at src/libcore/panicking.rs:96
  15: <vhost_user_fs::VhostUserFsBackend<F> as vhost_user_backend::VhostUserBackend>::handle_event
  16: vhost_user_backend::VringWorker::run
```

bjzhjing commented:
> Pretty clear in the output to me, search for "panic":
> 
> https://cloud-hypervisor-jenkins.westus.cloudapp.azure.com/job/cloud-hypervisor/job/PR-785/6/consoleFull
> 
> ```
> thread 'vring_worker' panicked at 'index out of bounds: the len is 1 but the index is 1', src/bin/vhost_user_fs.rs:173:33
> stack backtrace:
>    0: backtrace::backtrace::libunwind::trace
>              at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/libunwind.rs:88
>    1: backtrace::backtrace::trace_unsynchronized
>              at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/mod.rs:66
>    2: std::sys_common::backtrace::_print_fmt
>              at src/libstd/sys_common/backtrace.rs:77
>    3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
>              at src/libstd/sys_common/backtrace.rs:61
>    4: core::fmt::write
>              at src/libcore/fmt/mod.rs:1028
>    5: std::io::Write::write_fmt
>              at src/libstd/io/mod.rs:1412
>    6: std::sys_common::backtrace::_print
>              at src/libstd/sys_common/backtrace.rs:65
>    7: std::sys_common::backtrace::print
>              at src/libstd/sys_common/backtrace.rs:50
>    8: std::panicking::default_hook::{{closure}}
>              at src/libstd/panicking.rs:188
>    9: std::panicking::default_hook
>              at src/libstd/panicking.rs:205
>   10: std::panicking::rust_panic_with_hook
>              at src/libstd/panicking.rs:464
>   11: std::panicking::continue_panic_fmt
>              at src/libstd/panicking.rs:373
>   12: rust_begin_unwind
>              at src/libstd/panicking.rs:302
>   13: core::panicking::panic_fmt
>              at src/libcore/panicking.rs:139
>   14: core::panicking::panic_bounds_check
>              at src/libcore/panicking.rs:96
>   15: <vhost_user_fs::VhostUserFsBackend<F> as vhost_user_backend::VhostUserBackend>::handle_event
>   16: vhost_user_backend::VringWorker::run
> ```
Thanks! I've addressed all the comments above.

rbradford commented on vhost_user_net/src/lib.rs:
```diff
@@ -93,15 +99,14 @@ impl std::convert::From<Error> for std::io::Error {
     num_queues: usize,
     queue_size: u16,
+    vrings: Option<Vec<Arc<RwLock<Vring>>>>,
+    worker_thread: Option<Vec<thread::JoinHandle<VhostUserBackendResult<()>>>>,
```
Plural. Does it even need to be an `Option<T>` now it's a `Vec` ?

bjzhjing replied:
worker_thread in vhost_user_net will be created after memory update, before that,  it should be initialized with None.

rbradford commented on src/bin/vhost_user_fs.rs:
```diff
@@ -184,6 +184,22 @@ impl<F: FileSystem + Send + Sync + 'static> VhostUserBackend for VhostUserFsBack
         self.vu_req = Some(vu_req);
     }
+
+    fn get_epoll_number(&self) -> usize {
```
s/get_epoll_number/num_epolls/g

bjzhjing replied:
update is done!

rbradford commented on src/bin/vhost_user_fs.rs:
```diff
@@ -184,6 +184,22 @@ impl<F: FileSystem + Send + Sync + 'static> VhostUserBackend for VhostUserFsBack
+        self.num_queues() / 2
+    }
+
+    fn get_vring_subset(
```
s/get_vring_subset/vring_subset/

bjzhjing replied:
update done!

rbradford commented on vhost_user_backend/src/lib.rs:
```diff
@@ -113,6 +115,32 @@ pub trait VhostUserBackend: Send + Sync + 'static {
+
+    /// Get backend type
+    fn get_type(&self) -> u32 {
+        0
```
In Rust style accessors should not really start with get_

rbradford replied:
Drop get_type() it's a bad hack.

bjzhjing replied:
Why do you think it is a bad hack?

bjzhjing replied:
Anyway, this one is removed.

rbradford commented on vhost_user_backend/src/lib.rs:
```diff
@@ -420,6 +455,13 @@ impl VringWorker {
+    /// Get epoll_fd. If vhost-user backend would like to run their specified
+    /// handler thread, it requires to listen to events on the epoll fd
+    /// defined in vring_worker since virtqueues register their events there.
+    pub fn get_epoll_fd(&self) -> RawFd {
```
Ditto.

bjzhjing replied:
Removed.

rbradford commented on vhost_user_backend/src/lib.rs:
```diff
@@ -841,7 +880,9 @@ impl<S: VhostUserBackend> VhostUserSlaveReqHandler for VhostUserHandler<S> {
     fn drop(&mut self) {
-        if let Some(thread) = self.worker_thread.take() {
+        let thread_number = self.worker_threads.len();
+        for _ in 0..thread_number {
```
Don't iterate over the count just over the vector using .drain()

rbradford commented on vhost_user_block/src/lib.rs:
```diff
@@ -270,8 +270,8 @@ impl VhostUserBackend for VhostUserBlkBackend {
         match device_event {
-            q if device_event < self.config.num_queues => {
-                let mut vring = vrings[q as usize].write().unwrap();
+            _q if device_event < self.config.num_queues => {
```
Why this change?

bjzhjing replied:
> Please split this commit into smaller changes:
> 
> * Introduce infrastructure into vhost-user-backend to allow the backends to control their worker thread
> * Arc
> * Port -net to use it
> 
> With the exception of if you change function names I don't think there should be changes to -block and -fs.

The three items above are the current existing commits, you mean to split them into smaller ones? While there are new methods added in VhostUserBackend to acquire backends' information to support multiple threads, since -block and -fs have different behavior on that, changes are required to provide concrete implementation respectively. Let me know if you still think the changes is not needed.

rbradford commented on vhost_user_net/src/lib.rs:
```diff
@@ -94,7 +94,7 @@ impl std::convert::From<Error> for std::io::Error {
 pub struct VhostUserNetBackend {
     mem: Option<GuestMemoryMmap>,
-    vring_worker: Option<Arc<VringWorker>>,
+    vring_worker: Option<Vec<Arc<VringWorker>>>,
```
Plural

bjzhjing replied:
No change it since I do not see the benefit. Is there any potential issue if not remove Option?

rbradford commented on vhost_user_backend/src/lib.rs:
```diff
@@ -537,17 +562,21 @@ impl<S: VhostUserBackend> VhostUserHandler<S> {
-                    .spawn(move || vring_worker.run(vring_handler))
-                    .map_err(VhostUserHandlerError::SpawnVringWorker)?,
-            );
+            if VirtioDeviceType::from(backend.read().unwrap().get_type())
```
Wrong way to do it, instead query the backend for it's workers and use them if provided.

bjzhjing replied:
Remove the type query, but we could not get the worker_thread from backends at this moment, since it will be created after memory is update.

rbradford commented on vm-virtio/src/lib.rs:
```diff
@@ -62,11 +62,11 @@ const VIRTIO_F_IOMMU_PLATFORM: u32 = 33;
 
 // Types taken from linux/virtio_ids.h
-#[derive(Copy, Clone)]
+#[derive(Copy, Clone, PartialEq)]
```
Drop this change.

bjzhjing replied:
Drop it.

rbradford commented:
Also you need to add integration test with non-MQ and MQ testing.

bjzhjing commented:
> Also you need to add integration test with non-MQ and MQ testing.

There are separately test cases for non-MQ and MQ currently.

sboeuf commented on vhost_user_net/src/lib.rs:
```diff
@@ -151,42 +143,193 @@ impl VhostUserNetBackend {
         Self::new_with_tap(taps, num_queues, queue_size)
     }
 
+    pub fn set_vring_worker(&mut self, vring_worker: Option<Vec<Arc<VringWorker>>>) {
```
Looks like you moved some code around on this commit, which makes the review harder. Could you make sure you only make the changes needed, and do the moving into a different commit?

bjzhjing replied:
Oh, I'm sorry for making review hard. Since NetEpollHandler is added in vhost_user_net backend, I re-adjust the code layout, some functions are moved around. There are implementations for the three parts currently: VhostUserNetBackend, VhostUserBackend  and NetEpollHandler. For all the changes in this commit aim to enable NetEpollHandler actually, I think we could keep them as is.

sboeuf commented on vhost_user_backend/src/lib.rs:
```diff
@@ -113,6 +113,40 @@ pub trait VhostUserBackend: Send + Sync + 'static {
+    ///         since backend has created by itself.
+    fn is_worker_thread_required(&self) -> bool {
+        true
+    }
```
What's the reason why we would want the backend library not to create the worker thread? That would defeat the purpose of this crate if the backend implementation was creating the worker thread on its own.

sboeuf replied:
Also, this would simplify the code.

bjzhjing replied:
> What's the reason why we would want the backend library not to create the worker thread? That would defeat the purpose of this crate if the backend implementation was creating the worker thread on its own.


For vhost-user-net backend, the worker threads need to access RxVirtio/TxVirtio and change the buffer there, those data structure defined in backend now, there will be concurrent access which will lead to performance penalty, that's why there is no performance improvement in vhost-user-net backend test although multiple thread is added in vhost-user backend lib code. Thus, each thread should access it's own data structures to avoid race condition, but those structures are vhost-user-net backend specified, it's not reasonable to cover them in vhost-user lib code. To resolve this issue, methods are added to let backends create their own threads.

sboeuf commented on vhost_user_backend/src/lib.rs:
```diff
@@ -113,6 +113,40 @@ pub trait VhostUserBackend: Send + Sync + 'static {
+    /// vring events.
+    fn vring_worker_thread(
+        &self,
+    ) -> Option<Vec<thread::JoinHandle<result::Result<(), io::Error>>>> {
```
Maybe this can be removed if we don't think there's a use case where the backend implementation would create its own worker thread.

rbradford commented on vhost_user_block/src/lib.rs:
```diff
@@ -270,8 +270,8 @@ impl VhostUserBackend for VhostUserBlkBackend {
-            q if device_event < self.config.num_queues => {
-                let mut vring = vrings[q as usize].write().unwrap();
+            q if q < self.config.num_queues => {
+                let mut vring = vrings[0].write().unwrap();
```
Why are you checking this value and ignoring it?

bjzhjing replied:
device_event is the event index registered on epoll_fd, it's equal to the original queue index. However, while create multiple worker threads in vhost_user lib, for vhost_user_blk, it's assumed that one thread will handle one queue, the queue index in handle_event should always be zero, but I think we could keep the checking on device_event here to ensure it's a validate event.

bjzhjing commented:
@rbradford @sboeuf @sameo This PR is rebased on #887 

Run the test with kernel v5.6-rc6.

> 2 queue pairs, run 2 iperf3 sessions, 2vcpus
```
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-30.00  sec  77.7 GBytes  22.3 Gbits/sec    0             sender
[  5]   0.00-30.00  sec  77.7 GBytes  22.3 Gbits/sec                  receiver
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-30.00  sec  77.7 GBytes  22.3 Gbits/sec    0             sender
[  5]   0.00-30.01  sec  77.7 GBytes  22.3 Gbits/sec                  receiver
```

> 1 queue pairs, 2 iperf3 sessions, 2vcpus
```
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-30.00  sec  55.4 GBytes  15.9 Gbits/sec    0             sender
[  5]   0.00-30.00  sec  55.4 GBytes  15.9 Gbits/sec                  receiver
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-30.00  sec  55.8 GBytes  16.0 Gbits/sec    0             sender
[  5]   0.00-30.00  sec  55.8 GBytes  16.0 Gbits/sec                  receiver

```
throughput is improved ~39%.

sboeuf commented:
@bjzhjing nice numbers :)
FYI, I'm looking at the vhost_user_backend crate because I think we can re-arrange it so that multiqueue/multithread can be supported through the crate. That would prevent from the duplication in each backend.
I'll keep you posted on that, and if I can't find a way, we'll fallback onto your solution.
Thanks for working on this :)

bjzhjing commented:
> @bjzhjing nice numbers :)
> FYI, I'm looking at the vhost_user_backend crate because I think we can re-arrange it so that multiqueue/multithread can be supported through the crate. That would prevent from the duplication in each backend.
> I'll keep you posted on that, and if I can't find a way, we'll fallback onto your solution.
> Thanks for working on this :)

Ok. Please ensure that net backend with mq enabled will gain performance improvement with your new changes, that's why we separate the thread management from lib code :-)

bjzhjing commented:
Close this one for #1012 has done the same thing.
